api/src/Serializer/Normalizer/UserNormalizer.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace App\Serializer\Normalizer;
+
+use App\Entity\User;
+use App\Metadata\Resource\Factory\UriTemplateFactory;
+use Rize\UriTemplate;
+use Symfony\Component\Serializer\Exception\ExceptionInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+use Symfony\Component\Serializer\SerializerAwareInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+
+class UserNormalizer implements NormalizerInterface, SerializerAwareInterface {
+    public function __construct(
+        private NormalizerInterface $decorated,
+        private UriTemplate $uriTemplate,
+        private UriTemplateFactory $uriTemplateFactory,
+    ) {
+    }
+
+    public function supportsNormalization($data, string $format = null) {
+        return $this->decorated->supportsNormalization($data, $format);
```

Comment:
Eigentlich wäre die Idee in API Platform dass du in diesem Fall hier nicht dekorierst, sondern "normal" Composition anwendest: https://api-platform.com/docs/core/serialization/#changing-the-serialization-context-on-a-per-item-basis

In `supportsNormalization` würde dann der `$data instanceof User` Test stehen, plus ein Rekursions-Schutz. Und dafür würde `normalize` wirklich nur noch laufen, wenn `$object` ein `User` ist. Ein Beispiel dafür ist der `InputFilterDenormalizer`.

Das würde eigentlich auf diesen Normalizer sowie den ContentTypeNormalizer zutreffen, ich nehme an von dem hast du kopiert. Bin aber selber unsicher ob diese Unterscheidung von Dekoration und Komposition wirklich wichtig ist.

Reply:
Da bin ich überfragt.
Wir haben momentan noch keinen Normalizer mit Composition.

Die Beispiele zeigen wie man 
eine normalization Group aufgrund der Rolle hinzufügt für alle Items -> Decorated
den context auf per item basis ändert -> Composition

Mein Fall wird da nicht abgedeckt, und eventuell könnte man beides mit beiden varianten implementieren...?

Mein Fall wäre eher das Beispiel https://api-platform.com/docs/core/serialization/#decorating-a-serializer-and-adding-extra-data
(Wobei dafür könnte man auch unterschiedliche normalization_contexts verwenden)
Aber ich möchte ja unbedingt einen berechneten Link einfügen.
(Sonst müssten wir die property definitionen kopieren, oder 2 Entities mit inheritance haben, oder anderes komisches Zeugs).

Ich würde sagen, solange wir nicht wissen, wann wir Composition und wann Dekoration verwenden sollen, würde ich es überall möglichst gleich machen.

Reply:
Hast du das eventuell so gemeint. (Unterschied zwischen generellen normalizern für alle Ressourcen, und solchen, die nur für eine Resource gelten):

![grafik](https://user-images.githubusercontent.com/1506818/142768229-c3ca8f49-85dc-4dd7-b873-6f38a84699d4.png)
http://www.plantuml.com/plantuml/uml/bPDVI_DG3CRl-od2tejNBzhT3v6DYJgG8NFU2k4qdGTFSqgQuH_nkzkMuTdOcRMzJ9xaTvuadTQAejlI9V-VLQjwCXvN3hcDPCQNO-EmhjFSuaQmJ18XeyWRHp2w8IP11rQf1FPIehCV9FK8FXDeFlPAeBu2Ny0jkaKZM_we9d0jleHPPU7UeHPDeYlgSCVoNSklP4-qegQGyhbtZenQptUMNuvAsY-545wmaXHe278oNjeEnwtdLipMeQoe824sT4NQCdWJVzbWp130HPRrUUrwnYi7a9BcSVjB2vdPAmgj1Qt6zjRDBAhU3UTVT32y1puuh20-gy-XGS3SinBh-hsYxhh2eaBYzsDPqfi-L6GikjCr0zSOfxTJxyNrEdwiIGwjjf4_UuoAXvaBm8YjG0X-eWVmbARnCPsKTzm-NPPTzluJoPGutvRk6m00

Reply:
Ja, genau um diese Unterscheidung geht es. Ich hatte diese Patterns bisher so in der [API Platform Dokumentation](https://api-platform.com/docs/core/serialization/#changing-the-serialization-context-on-a-per-item-basis) und [SymfonyCasts](https://symfonycasts.com/screencast/api-platform-security/normalizer-aware) gesehen, und darum versucht zu übernehmen, bin aber selber nicht sicher was überhaupt die Begründung für die Unterscheidung ist.

So oder so fände ich es leicht schöner, wenn das grosse `if ($object instanceof User) {` von der Methode `normalize` in die Methode `supports` verschoben werden könnte. Das hat überhaupt meinen Kommentar ausgelöst.

Nebenbei: Ich glaube ich habe das Pattern selber nicht korrekt angewendet: Der `InputFilterNormalizer` den ich für die InputFilter `Trim`, `CleanHTML` etc. geschrieben habe greift allgemein bei allen Entities ein, verwendet aber Composition statt Decoration. Und der ContentTypeNormalizer von usu verwendet aktuell Decoration statt Composition. Trotzdem funktionierts bisher nicht schlecht, was auch ein Argument dafür ist, überall nur eine Variante zu verwenden.

Reply:
Wir lassens mal so bis es das nächste Mal wieder auftritt